src/cmd/create.go

```diff
@@ -710,19 +717,48 @@ func pullImage(image, release, authFile string) (bool, error) {
 		shouldPullImage = true
 	}
 
+	imageSizeChan := make(chan string)
+	confirmationChan := make(chan bool)
+	var userResponded bool
+
 	if promptForDownload {
 		fmt.Println("Image required to create toolbox container.")
 
-		var prompt string
+		prompt := fmt.Sprintf("Download %s (...)? [y/N]:", imageFull)
 
-		if imageSize, err := getImageSizeFromRegistry(imageFull); err != nil {
-			logrus.Debugf("Getting image size failed: %s", err)
-			prompt = fmt.Sprintf("Download %s? [y/N]:", imageFull)
-		} else {
-			prompt = fmt.Sprintf("Download %s (%s)? [y/N]:", imageFull, imageSize)
-		}
+		askForConfirmationAsync(prompt, confirmationChan)
 
-		shouldPullImage = askForConfirmation(prompt)
+		// Launch a goroutine to call getImageSizeFromRegistry and send the size to the channel
+		go func() {
+			imageSize, err := getImageSizeFromRegistry(imageFull)
+			if err != nil {
+				logrus.Debugf("Getting image size failed: %s", err)
+				imageSize = ""
+			}
+			imageSizeChan <- imageSize
+		}()
+
+		var imageSize string
+		select {
+		case imageSize = <-imageSizeChan:
+			if imageSize != "" {
+				prompt = fmt.Sprintf("\rDownload %s (%s)? [y/N]:", imageFull, imageSize)
+			} else {
+				prompt = fmt.Sprintf("\rDownload %s? [y/N]:", imageFull)
+			}
+			fmt.Printf("%s", prompt)
+			userResponded = false
+			for !userResponded {
+				select {
+				case shouldPullImage = <-confirmationChan:
+					userResponded = true
+				case <-time.After(time.Second):
+					break
+				}
+			}
+		case shouldPullImage = <-confirmationChan:
+			break
+		}
```

Comment:
Just like @debarshiray pointed out, there's a race condition here. Additionally, we have an unresolved design decision. First, about the race.

Rishi already summed it up well, so I'll just highlight some concrete cases when things can go wrong:

1. The first example is simple and is more of a logical error. The dialog is first printed by `askForConfirmation()` and once the image size is received, we call `fmt.Printf()` outside of it. But the prompt in `askForConfirmation()` does not fail fast and is forgiving when the user provides unknown input. Hence, you can end up with output like:
  
  ```
  ; toolbox create -r 40
  Image required to create toolbox container.
  Download registry.fedoraproject.org/fedora-toolbox:40 (1.014GB)? [y/N]:foo
  Download registry.fedoraproject.org/fedora-toolbox:40 (...)? [y/N]: foo
  Download registry.fedoraproject.org/fedora-toolbox:40 (...)? [y/N]: 
  ```
  
  The `1.014GB` is replaced by `...` because the value was only printed but not kept around and definitely not communicated to `askForConfirmation()` which still has the old prompt in it. Thus, the value is lost.

2. The second case is a bit harder to reproduce but in the current implementation is possible. Let the scenario play out like this:

  ```
start askForConfirmation()
   \
    \            return getImageSizeFromRegistry()  
     \                \
  ---------------------------------------------------------------------
         /                   /                  /
        /                print new prompt    print old prompt
       /
  start getImageSizeFromRegistry()
  ```

This scenario is extremely unlikely but currently can happen due to two unsynchronized interactions with the prompt.

Both issues are a reason to create a fully async `askForConfirmation()` function which handles updates to the prompt in a synchronized and predictable manner.

Reply:
Thanks for reviewing @HarryMichal ! I'll continue working on it and try to achieve all the requirements.